### PR TITLE
[On Hold Until July 2021] [Experimental] PinchSliderBox + PinchSpring

### PR DESCRIPTION
**Pull request by Cameron-Micka**
## Overview

This pull request introduces two new experimental UX components that have been requested by our partners and designers.

**PinchSliderBox**: A control that automatically spawns PinchSliders to non-uniformly scale the target of a [TransformScaleHandler](https://microsoft.github.io/MixedRealityToolkit-Unity/api/Microsoft.MixedReality.Toolkit.UI.TransformScaleHandler.html).

PinchSliderBox

**PinchSpring**: Simulates a point mass on a spring to provide spring like interactions.

PinchSpring

All of these components have been combined together to simulate a "fixed axis clipping box" using the [ClippingBox ](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/README_MRTKStandardShader.html#primitive-clipping)primitive. Use the fixed axis clipping box to peer inside a human heart in the PinchSliderBoxExamples scene. _(Happy Halloween!)_

brain

Note, these controls are currently [experimental ](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html)and still being reviewed by our design team.

## Changes
- Fixes: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5803

This pull request also provides a opt-in fix for https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5975 by providing a **EventSystemExtensions.ExecuteHierarchyUpward(…)** method to bubble input events up.

## Verification
Please try the ClippingExamples scene.

>
> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

### CODE REVIEW

**Comment by Cameron-Micka** [comment]
I will start making those tests/changes now. Thank you for the feedback!

**Comment by julenka** [comment]
Also, the GIF is awesome!

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/TransformScaleHandler.cs
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/TransformScaleHandler.cs]: I think this, along with the ScaleMaximum, need comments.

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/TransformScaleHandler.cs
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/TransformScaleHandler.cs]: Also needs comment for docfx.

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityEventHandlers.cs
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityEventHandlers.cs]: It looks like you don't use this anywhere, did you mean to add this file? Perhaps it was added by accident?

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Oh, I see, it looks like you moved all of these into their own file. Any reason for this? Just to make this file shorter? I'm a bit wary to see a big change like this in an experimental feature, but if it's just moving code around I suppose it's okay.

File: Assets/MixedRealityToolkit.Examples/Demos/StandardShader/Scenes/ClippingExamples.unity
Comment (julenka) [inline Assets/MixedRealityToolkit.Examples/Demos/StandardShader/Scenes/ClippingExamples.unity]: Instead of putting an experimental feature in an example, could you instead create a new scene under examples/experimental/pinchboxslider/Clippingbrain.unity, or something like this? Otherwise people may not realize that the pinch slider box is still experimental.

Comment (julenka) [reply Assets/MixedRealityToolkit.Examples/Demos/StandardShader/Scenes/ClippingExamples.unity]: I believe you used a similar pattern for the hand menu

**Comment by cre8ivepark** [comment]
I did small adjustments to the content size and position. I also added some polish to the ClippingExamples scene.

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs
Comment (keveleigh) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: It would be great if we could get this working without flipping these to `public`. These definitions should really stay private to the input system and not directly referenced outside it.

Comment (keveleigh) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: It would be great if we could get this working without flipping these to `public`. These definitions should really stay private to the input system and not directly referenced outside it.

Comment (wiwei) [reply Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: +1

Comment (keveleigh) [reply Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: With Cameron out at the moment, I'm looking into this

Comment (Cameron-Micka) [reply Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Thank you @keveleigh, I just made you a collaborator.

Comment (Cameron-Micka) [reply Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Hey @keveleigh, do you have any thoughts on how you would like these handlers encapsulated? I can make the changes needed if you are busy with other tasks.

Comment (julenka) [reply Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Perhaps we could modify [MixedRealityInputSystem.DispatchEventToObjectFocusedByPointer](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs#L540) to allow for the [event capture phase](https://javascript.info/bubbling-and-capturing#capturing)?

Comment (Cameron-Micka) [reply Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: I'm not familiar with that method. But, it looks similar to where I'm using handles in: https://github.com/Cameron-Micka/MixedRealityToolkit-Unity/blob/44fb0bd836e18e58024f35181e8c299a8cacfa00/Assets/MixedRealityToolkit/Extensions/EventSystemExtensions.cs#L75

**Comment by davidkline-ms** [comment]
@Cameron-Micka, can you resolve the merge conflicts. Please note that the folder structure changed in the dev branch since this PR was opened.

@keveleigh, has your feedback been addressed ?

**Comment by keveleigh** [comment]
> has your feedback been addressed?

I don't think so directly yet, but #7550 should help with that, to take bubbling events further up the hierarchy largely out of the hands of these scripts.

**Comment by Cameron-Micka** [comment]
I can fix the merge conflicts real quick again. @keveleigh are you able to make the changes needed once #7550 goes in?

**Comment by Cameron-Micka** [comment]
Just fixed up all the merge conflicts. 👍

**Comment by keveleigh** [comment]
Marking as "do not merge". Currently blocked on #7550. I'll update things once that goes in.

**Comment by Cameron-Micka** [comment]
> Marking as "do not merge". Currently blocked on #7550. I'll update things once that goes in.

Thank you!

**Comment by Cameron-Micka** [comment]
@keveleigh it looks like #7550 was merged in, is there anything else you need from me?

**Comment by keveleigh** [comment]
@Cameron-Micka Right now it's still in a feature branch: 
image
I was waiting on it to come back into mrtk_dev before moving forward.

**Comment by Cameron-Micka** [comment]
@keveleigh oh, I missed that detail. Thank you!

**Comment by Cameron-Micka** [comment]
Bump (to avoid stale-ness)

**Comment by RogPodge** [comment]
Refreshed with main, but noticed that it's still blocked by this feature branch getting merged: https://github.com/microsoft/MixedRealityToolkit-Unity/tree/feature/event_propagation

Leaving it up for now, but depending on the other branch's ultimate status, may end up closing next week.

**Comment by RogPodge** [comment]
Making a note here:

The branch that this PR depends on: https://github.com/microsoft/MixedRealityToolkit-Unity/tree/feature/event_propagation is slated to be revisited in July, as it currently doesn't support some event types.

Once the above branch gets merged into main, we should be clear on finally merging this PR in

**Review by Jinky100 — APPROVED** [review]

**Comment by Zee2** [comment]
`PinchSpring` functionality may already be entirely covered by the (still experimental) Elastics system. To prevent duplication of code, the pinch spring behavior in this PR should probably be rewritten to utilize the elastics functions.

**Comment by Cameron-Micka** [comment]
@Zee2 I agree the spring behavior should be re-written to use the new elastics system. I believe this PR is so old at this point that it predates that system by a few months. :D 

@cre8ivepark, since this PR is getting pretty old. I will leave it up to you discrection if this is something we still want in MRTK-Unity. If not, please feel free to close this PR.

**Comment by RogPodge** [comment]
@Zee2 can you check to see how much of this PR is covered by the elastics system you're working on and take the appropriate actions to port over the relevant components? It would be a shame to have this work disappear because I think it looks pretty cool, and was basically completed until we started overloading it with event propagation. 

Regarding event propagation, I think it's still something that will be important for some of our UI controls, but I think that we should be able to handle the bulk of it through Unity's UI systems as we align more closely with that. I'd really hate to lose out on the new visual components of this PR, so I think it's ok to merge in those changes once we strip out the extra event propagation things.

File: Assets/MRTK/Services/InputSystem/MixedRealityInputSystem.cs
Comment (keveleigh) [inline Assets/MRTK/Services/InputSystem/MixedRealityInputSystem.cs:12]: We may want to throw this into an AssemblyInfo.cs for discoverability. You'll likely need to create one for this assembly, but the pipeline patching step should adjust.

File: Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs:1]: nit: it's been so long, we can update these license headers to the new style

File: Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs:16]: We probably want to adjust this to the correct format for rendering in VS. Our docs generation will convert it.

```suggestion
    /// The <see cref="MinMaxScaleConstraint"/> is 
```

File: Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs:213]:
```suggestion
        /// made the parent of the <see cref="MinMaxScaleConstraint"/> TargetTransform.
```

File: Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs
Comment (keveleigh) [inline Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs:253]:
```suggestion
        /// <see cref="MinMaxScaleConstraint"/>'s TargetTransform's parent.
```

File: Assets/MRTK/Services/InputSystem/MixedRealityInputSystem.cs
Comment (RogPodge) [reply Assets/MRTK/Services/InputSystem/MixedRealityInputSystem.cs]: Updated!

File: Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs
Comment (RogPodge) [reply Assets/MRTK/SDK/Experimental/PinchSliderBox/PinchSliderBox.cs:1]: Updated!

**Pull request merged**
